### PR TITLE
Remove Scheduler.detach

### PR DESCRIPTION
**Pull request by rgrinberg**
Implement task detaching entirely using fibers.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Adding reviewers based on their interest in fibers :)

**Comment by rgrinberg** [comment]
How embarrassing, `cancel_timer` was completely broken. This PR fixes this as well.

**Comment by rgrinberg** [comment]
Okay, this is ready. Would appreciate some testing @ulugbekna @mnxn

**Comment by ulugbekna** [comment]
Thanks for the PRs you've been sending lately! :-) 

It seems to work on my machine. I will test more. I can also review it tomorrow (as much as I understand fibers), if you don't mind waiting. 

After first two installs, ocamllsp was crashing with the error below, but I may just had forgotten to `eval $(opam env)`. Mentioning it just in case. 

```
ocamllsp: internal error, uncaught exception:
          File "ocaml-lsp-server/vendor/merlin/src/ocaml/utils/410/load_path.ml", line 83, characters 2-8: Assertion failed
          Raised at file "ocaml-lsp-server/vendor/merlin/src/ocaml/utils/410/load_path.ml", line 83, characters 2-46
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/persistent_env.ml", line 42, characters 12-53
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/persistent_env.ml", line 257, characters 16-58
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/persistent_env.ml" (inlined), line 304, characters 6-45
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/env.ml", line 840, characters 2-86
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/env.ml", line 2454, characters 16-31
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/env.ml", line 2590, characters 23-71
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 85, characters 13-41
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 268, characters 8-39
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 307, characters 14-49
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 387, characters 17-42
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 395, characters 16-37
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 395, characters 16-37
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 861, characters 25-44
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/printtyp.ml", line 1003, characters 23-46
          Called from file "format.ml", line 1265, characters 32-48
          Called from file "format.ml", line 1252, characters 4-20
          Called from file "format.ml", line 1252, characters 4-20
          Called from file "format.ml", line 1262, characters 32-48
          Called from file "format.ml", line 1262, characters 32-48
          Called from file "format.ml", line 1264, characters 32-48
          Called from file "format.ml", line 1262, characters 32-48
          Called from file "format.ml", line 1262, characters 32-48
          Called from file "format.ml", line 1265, characters 32-48
          Called from file "format.ml", line 1252, characters 4-20
          Called from file "format.ml", line 1313, characters 16-34
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/typing/410/typecore.ml", line 5164, characters 17-308
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/parsing/410/location.ml" (inlined), line 527, characters 24-35
          Called from file "ocaml-lsp-server/vendor/merlin/src/ocaml/parsing/410/location.ml" (inlined), line 528, characters 33-51
          Called from file "ocaml-lsp-server/src/ocaml_lsp_server.ml", line 133, characters 18-59
          Called from file "list.ml", line 103, characters 22-25
          Called from file "vendor/stdune/list.ml" (inlined), line 5, characters 19-33
          Called from file "ocaml-lsp-server/src/ocaml_lsp_server.ml", line 124, characters 12-590
          Called from file "vendor/fiber/fiber.ml", line 218, characters 36-41
          Called from file "vendor/fiber/fiber.ml", line 127, characters 18-21
          Re-raised at file "exn.cppo.ml" (inlined), line 50, characters 27-56
          Called from file "vendor/stdune/exn_with_backtrace.ml", line 13, characters 33-71
          Called from file "vendor/fiber/fiber.ml", line 143, characters 6-16
          Called from file "queue.ml", line 121, characters 6-15
          Called from file "vendor/fiber/fiber.ml", line 596, characters 10-33
          Called from file "exn.cppo.ml", line 13, characters 8-11
          Re-raised at file "exn.cppo.ml", line 19, characters 4-11
          Called from file "fiber-unix/src/scheduler.ml", line 377, characters 12-29
          Called from file "ocaml-lsp-server/src/ocaml_lsp_server.ml", line 774, characters 2-417
          Called from file "lsp/src/logger.ml", line 155, characters 8-12
          Re-raised at file "lsp/src/logger.ml", line 161, characters 4-19
          Called from file "lsp/src/logger.ml", line 187, characters 10-34
          Re-raised at file "lsp/src/logger.ml", line 193, characters 6-21
          Called from file "ocaml-lsp-server/vendor/cmdliner/src/cmdliner_term.ml", line 25, characters 19-24
          Called from file "ocaml-lsp-server/vendor/cmdliner/src/cmdliner.ml", line 117, characters 32-39
[Error - 5:33:10 PM] Connection to server got closed. Server will not be restarted.
```

**Comment by rgrinberg** [comment]
Yeah, I'm also encountering this error. However, I'm wondering if it's really something that is caused by this PR, or it's something that we've been ignoring previously.

@trefis @voodoos does the stacktrace above tell you anything?

**Comment by rgrinberg** [comment]
I'm in no rush to merge. I've cherry-picked some unrelated fixes to master already.

**Comment by mnxn** [comment]
I haven't encountered that error. Seems to work fine on OCaml 4.11.1 for both macOS and Windows.

Does this happen to fix the bug that prevented sending requests inside the `on_notification` handler? Previously that always resulted in `Never` exceptions.

**Comment by rgrinberg** [comment]
> Does this happen to fix the bug that prevented sending requests inside the on_notification handler? Previously that always resulted in Never exceptions.

Yup.

File: fiber-unix/src/scheduler.ml [hunk 1]
```diff
@@ -128,15 +135,10 @@ end
...
-  }
-
 type t =
   { mutable events_pending : int
   ; events : event Queue.t
-  ; mutex : Mutex.t
+  ; events_mutex : Mutex.t
   ; time_mutex : Mutex.t
   ; event_ready : Condition.t
   ; timers_available : Condition.t
```
Comment (ulugbekna) [inline fiber-unix/src/scheduler.ml:144]: I have started working on the review: I have more-or-less covered new version of `Fiber` lib (with slight rewrite and comments, it's much easier to understand, and comments help a lot!) I will continue tomorrow. 

I noticed that we have several places where mutexes are interleaved with data, e.g., the snippet selected. I can think of an easy change that would make code easier to understand (by removing boilerplate-y record fields) and less error-prone (by keeping mutex associated with the data closer to the data):

```ocaml
type 'data mutexed = { v : 'data ; m : Mutex.t }
let with_mutex_locked t ~f =
    Mutex.lock t.m;
    let res = f t.v in 
    Mutex.unlock t.m;
    res 
```

Comment (rgrinberg) [reply fiber-unix/src/scheduler.ml:144]: Thanks for the review. I'll add some comments.

> I noticed that we have several places where mutexes are interleaved with data, e.g., the snippet selected. I can think of an easy change that would make code easier to understand (by removing boilerplate-y record fields) and less error-prone (by keeping mutex associated with the data closer to the data):

It's worth experimenting with. Even better would be to replace things with mvars whenever possible.

**Comment by ulugbekna** [comment]
Oh, I was planning to do further reviewing today, but ok. I hope the
crashes we get aren't caused by this PR.

**Comment by rgrinberg** [comment]
Oops. Feel free to keep reviewing. We aren't anywhere near a release, and nothing in this implementation is final.

I can't see how the crashes can be related to this PR. however I have some ideas where we might be using merlin wrong.

**Comment by voodoos** [comment]
> Yeah, I'm also encountering this error. However, I'm wondering if it's really something that is caused by this PR, or it's something that we've been ignoring previously.
> 
> @trefis @voodoos does the stacktrace above tell you anything?

Do you have some steps to reproduce this error ?

Maybe @trefis or @let-def will know more, but I did not encountered this stacktrace before.

**Pull request merged**